### PR TITLE
less strict size check in scatter

### LINKED ISSUE #63265: torch.scatter_: support index.size(d) > src.size(d)
## 🚀 Feature
Teach [torch.Tensor.scatter_](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html) to handle `index.size(d) > src.size(d)`.

## Motivation

Currently, `torch.Tensor.scatter_` requires `index.size(d) <= src.size(d)` for all dimensions `d`, _unless_ `src` is float-valued. This constraint seems artificial.

```
import math
import torch

device = 'cuda'  # fails if device == 'cpu' (cf. #61854)

dim = 1
shape_src = 3, 5
shape_idx = 3, 7

index = torch.empty(shape_idx, dtype=torch.long, device=device).random_(0, shape_src[dim])

# succeeds
output = torch.zeros(shape_src, device=device)
output.scatter_(dim, index, 1.0, reduce='add')
print(output)

# desired usage
# output = torch.zeros(shape_src, device=device)
# values = torch.rand(shape_src, device=device)
# output.scatter_(dim, index, values, reduce='add')   # <-- failure
# print(output)
```

## Alternatives

An inefficient workaround is to chunk the index tensor:
```
output = torch.zeros(shape_src, device=device)
values = torch.ones(shape_src, device=device)
chunks = math.ceil(index.shape[dim] / output.shape[dim])
for index_chunk in torch.chunk(index, chunks, dim):
    output.scatter_(dim, index_chunk, values, reduce='add')
print(output)
```

### PR DESCRIPTION
**Pull request by bfgray3**
Fixes #63265

### CODE REVIEW

**Comment by nikitaved** [comment]
To allow such an improvement we have to modify the strided form of inputs tensor to be fed into a TensorIteretor. Currently, source is being restrided to match the size of an input, but we have to make sure that the restrided source and input is actually of shape:
```
src_restrided.shape[i] == min(src.shape[i], index[i])
```

Therefor, it is a bit more involved than changing the shape check. @bfgray3 , I could help you out with this one!

**Review by nikitaved — CHANGES_REQUESTED** [review]
We need to change inputs that are iterated over in a TensorIterator. See my comment above. And feel free to ask any questions regarding these modifications!

**Comment by bfgray3** [comment]
thanks for the pointer @nikitaved! i'll take a look over the next few days and let you know if i have questions.

**Comment by bfgray3** [comment]
@nikitaved am i looking in the right spot with the struct linked below for what needs to be changed?

https://github.com/pytorch/pytorch/blob/6db8f7a70920f91418078fe09477eed0b0adefdb/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L101

**Comment by nikitaved** [comment]
@bfgray3 , it is starting from here: https://github.com/pytorch/pytorch/blob/6db8f7a70920f91418078fe09477eed0b0adefdb/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L194.

So, what is there to be done?
1. remove the check between the src and the input.
2. compute the right shape and make sure that it is valid for the output.
3. modify the older tests (probably) and include some new ones that test the change.
4. update the documentation.

Something along these lines, I guess, is sufficient.

**Comment by bfgray3** [comment]
@nikitaved i'm very sorry, but i will be unable to complete this pull request.  i'll make a note in the issue.

**Comment by Sirish07** [comment]
> @bfgray3 , it is starting from here:
> 
> https://github.com/pytorch/pytorch/blob/6db8f7a70920f91418078fe09477eed0b0adefdb/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp#L194
> 
> So, what is there to be done?
> 
> 1. remove the check between the src and the input.
> 2. compute the right shape and make sure that it is valid for the output.
> 3. modify the older tests (probably) and include some new ones that test the change.
> 4. update the documentation.
> 
> Something along these lines, I guess, is sufficient.

Hey @nikitaved ,

I completed 1, 3, and 4 parts as mentioned above. In part 2, can you elaborate on what you meant by right shape? I would also like to know how src_restrided / index_restrided can be computed as mentioned above. I shall open a new PR regarding the same.

**Comment by nikitaved** [comment]
Hi, @Sirish07 . Have a look at https://github.com/pytorch/pytorch/pull/63452#issuecomment-909004767

**Comment by Sirish07** [comment]
Thanks @nikitaved!

> ```
> src_restrided.shape[i] == min(src.shape[i], index.shape[i])
> index_restrided.shape == src_restrided.shape // must be True
> // pass src_restrided and index_restrided to a TensorIterator
> ```
> We need to change inputs that are iterated over in a TensorIterator. See my comment above. And feel free to ask any questions regarding these modifications!

From what I understand, we need to compute src_restrided and index_restrided from src and index and pass them onto TensorIteratorConfig(). 

For eg: if we have src (dim = (3, 5)) and index (dim = (3, 7)), src_restrided (dim = (3,5)) can be computed as above (src_restrided = src). But, I am not sure about index_restrided. Can you tell me if I am going on the right path?
